### DISCUSSION TITLE
Jupyter Notebook Support

### DISCUSSION THREAD

**Discussion by amangalampalli** [community/community#10113] [Copilot Conversations]
While Visual Studio Code has a Jupyter Notebook extension, I tend to use the default Jupyter Notebook in the web browser as that seems more comfortable to me. Is there an extension for Github Copilot in the Jupyter Notebook planned for sometime in the future?

**Comment by BrunoGomesCoelho** [comment]
+1 for native JN (outside VS Code) support!

**Comment by krassowski** [comment]
Just as a note, since JupyterLab 4.1.0 and Jupyter Notebook 7.1.0 (released Feb 2024) both have an inline completer:
- https://jupyterlab.readthedocs.io/en/latest/user/completer.html#inline-completer
- https://jupyterlab.readthedocs.io/en/latest/extension/extension_points.html#inline-completer
- https://jupyterlab.readthedocs.io/en/latest/api/interfaces/completer.IInlineCompletionProvider.html

which is integrated with [jupyter-ai](https://github.com/jupyterlab/jupyter-ai) among others. If the Copilot team has any questions or request please do contact me (as a person who implemented the inline completion API).

Also see https://github.com/orgs/community/discussions/63345.

**Reply by jtpio** [reply]
For reference, I looked into integrating with some online providers to have copilot-like code completions in JupyterLab and Jupyter Notebook.

Using the inline completer API mentioned above to implement this functionality as a JupyterLab extension, which also work with Jupyter Notebook.

One of the providers is Codeium, which is integrated in this extension: https://github.com/jtpio/jupyterlab-codeium

[jupyterlab-codeium.webm](https://github.com/community/community/assets/591645/00f0333b-0f0f-40cc-a0f4-79d604d27e36)

This was also explored in another extension, based on the MistralAI models: https://github.com/jtpio/jupyterlab-codestral

[jupyterlab-codestral-demo-lite.webm](https://github.com/community/community/assets/591645/ce308f73-5405-4477-aa64-bb36ce317dc8)

Posting this here as examples, in case some folks would be interested in integrating GitHub Copilot in JupyterLab and Jupyter Notebook.
